### PR TITLE
Initial Plugin System Expansion

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Working towards #1055, #819.

##### What are the significant changes in functionality due to this change request?
Implements an initial structure for management of plugins.

- Extends plugin installer to allow installation of plugins in non-default locations
- Implements a plugin handler to perform actions related to plugin identification and location
- Implements a Plugin Factory for loading and providing entities from plugins
- Adds a plugin directory file for tracking installed plugins
- Provides initial implementation of plugin manual for guiding plugin developers
- Reworks library management in RAVEN to use config files
- Implements plugins extending required and optional libraries for various installations
- Implements Plugin.Entity format for library-style identification of plugin entities
- Extends `run_tests` to run `raven` tests first, then iterate over plugin tests

Additional improvements
- Provides examples of wildcard usage for Generic Code interface

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: doc/plugins_manual/introduction.tex [hunk 1]
```diff
@@ -0,0 +1,46 @@
+\section{Introduction}
+
+RAVEN (Risk Analysis Virtual ENvironment) [TODO cite] is a powerful tool for risk- and uncertainty-based stochastic analysis. TODO say more.
```
Comment (alfoa) [inline doc/plugins_manual/introduction.tex]: can you add the citation and comment out the "TODO say more"?

File: doc/plugins_manual/introduction.tex [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+Occasionally, the applications of RAVEN become sufficiently varied and complex to warrant a special application to a particular suite of problems. These special applications may introduce new physics, new templates, new RAVEN entities, or any combination of the above. In this instance, it's worth considering whether developing a Plugin could be beneficial to you and the community.
+
+\subsection{What are Plugins?}
+
+Plugins are defined the RAVEN Software Quality Assurance documentation [TODO CITE]. They are intended to allow development that does not fit in the scope of RAVENs mission to be completed in a way that is still captured under the umbrella of RAVEN-based tools.
```
Comment (alfoa) [inline doc/plugins_manual/introduction.tex]: citation

File: scripts/library_report [hunk 3]
```diff
@@ -38,9 +38,10 @@ try:
 except:
   print("Failed to find git version")
 sys.path.append(os.path.join(RAVEN_DIR, "scripts"))
-sys.path.insert(0,os.path.join(RAVEN_DIR, "framework", "utils"))
+#sys.path.insert(0,os.path.join(RAVEN_DIR, "framework", "utils"))
```
Comment (alfoa) [inline scripts/library_report]: remove commented code

Comment (PaulTalbot-INL) [reply scripts/library_report]: done

File: doc/plugins_manual/introduction.tex (hunk 2 above)
Comment (PaulTalbot-INL) [reply doc/plugins_manual/introduction.tex]: What's the citation for our sqa material?

File: doc/plugins_manual/introduction.tex (hunk 1 above)
Comment (PaulTalbot-INL) [reply doc/plugins_manual/introduction.tex]: done

File: doc/plugins_manual/introduction.tex (hunk 2 above)
Comment (PaulTalbot-INL) [reply doc/plugins_manual/introduction.tex]: I cited the user manual.

File: framework/Models/ModelPlugInFactory.py [hunk 4]
```diff
@@ -57,6 +57,8 @@ def registerSubtypes(typeName, typeDict):
   for name, subtype in typeDict.items():
     __interfaceDict[typeName][name] = subtype
     __knownTypes.append(name)
+  if len(__knownTypes) == 1:
+    raise RuntimeError()
```
Comment (alfoa) [inline framework/Models/ModelPlugInFactory.py]: remove

File: run_tests [hunk 5]
```diff
@@ -24,12 +28,19 @@ else
...
 for A in "$@"; do
     case $A in
         --skip-conda)
             SKIP_CONDA=0
             ;;
+        --plugins)
+            TEST_SET=1
+            ;;
+        --test_all)
+            TEST_SET=2
```
Comment (alfoa) [inline run_tests]: all should be the default

File: framework/Models/ModelPlugInFactory.py (hunk 4 above)
Comment (PaulTalbot-INL) [reply framework/Models/ModelPlugInFactory.py]: done

File: run_tests (hunk 5 above)
Comment (PaulTalbot-INL) [reply run_tests]: done, but this requires a change in CIVET not to run the plugins as required. PR is pending.

**Comment by PaulTalbot-INL** [comment]
Needs to rerun tests once https://github.com/idaholab/civet_recipes/pull/1499 is merged.

**Comment by alfoa** [comment]
all the machine are allowed to fail now?

**Comment by PaulTalbot-INL** [comment]
Should retest windows after https://github.com/idaholab/civet_recipes/pull/1500

**Comment by PaulTalbot-INL** [comment]
Not the machines, no; the individual plugins tests on each machine are allowed to fail, yes.

**Comment by alfoa** [comment]
@PaulTalbot-INL Unfortunately I can not merge your stuff... I do not have Write/Admin rights!

**Comment by PaulTalbot-INL** [comment]
The change control board not having write privileges is going to make it very difficult to progress.

I will restart civet on windows before COB today, so at least we can see the results.

**Comment by alfoa** [comment]
Pull Request is approved.

The Tests pass and the official plugins are failing. This is allowed and the developers of the plugin have been informed.

**Comment by alfoa** [comment]
PR Checklist passed

**Pull request merged**